textpipe/doc.py

```diff
@@ -226,6 +237,18 @@ def nwords(self):
         """
         return len(self.clean.split())
 
+    @property
+    def words(self, model_name=None):
+        """
+        Extract the text and character offsets of words from text
+
+        >>> doc = Doc('Test sentence for testing text.')
+        >>> doc.words
+        [(0, 4, 'Test'), (5, 13, 'sentence'), (14, 17, 'for'), (18, 25, 'testing'), (26, 30, 'text'), (30, 31, '.')]
```

Comment:
For both words and sentences, I feel like given both start and end index with the string is overkill (as `span.start_char + len(span.text) == span.end_char`).

Reply:
Good point! Our considerations were first that we want start + end offsets for quick slicing of the original document (and more generally in our pipeline we have opted for a offset-only representation for most of our representations and enrichements). We added the `text` because we figured this can be convenient in use-cases in general :-) (and it's free). 

Having them all three does introduce some redundancy, however, removing end offsets means losing the ability for quick slicing; but is an option, removing the text another, or, keeping it all in as we have now. 

We thought of slicing by using start_offset + next_start_offset-1 but not sure if that's always true? 

Anyone else have strong opinions or bright insights on the matter?

Reply:
I'd vote for killing `text` over end-offsets, btw, if it'd come to a vote ;-).

Reply:
I would vote for (text, start_char)-tuples; I would definitely expect at least a string from a `words` or `sentences` property.

Reply:
Maybe different use-cases deserve different properties; what about having a `sents` prop for text, and `sent_offsets` with char_start/char_end (or char_start/char_len) tuples? Same for `words`

Reply:
I think that it would make sense to combine sents/words with their start index; this includes all information and as we're actually detecting boundaries in strings it makes sense to also return the boundary indices with the text.

Reply:
On @anneschuth 's iterator point: I would prefer an iterator as it's also more in line with Spacy (Spacy generally is as lazy as possible).

Reply:
But we don't care about spacy! we care about textpipe :-). @anneschuth's remark seems to suggest there aren't iterators where there {sh,c}ould be iterators? I personally don't mind too much either way.

Reply:
Ok regarding the iterator bit, maybe think about the use cases? Advantage of iterator would be more efficient if we process larger texts. Are there use cases where we would not process the document in order? Otherwise, list might be handy... 
If we would opt for iterator, then should we return the iterator spacy provides, even though these iterators contain more elements than we currently use (for example POS tags of tokens in the word case)?
I don't really have a preference, for our use case a list would be fine, and it is consistent with textpipe not having iterators so far, but in terms of future thinking, iterator may be better.

Reply:
Sure, but that should be up to the user of the library, its trivial to convert to a list from an iterator, the other way around isn't possible.

Reply:
I don't think the choice if iterator vs list should determine what's in there. Thats a separate issue.

Reply:
Regarding having 2 separate properties (one for offset, one for text) I am not sure whether we should do that - there is a risk that if offset is applied to the wrong text e.g. raw vs clean for example, unexpected things happen. While if offset and text are returned together this can always be checked.

Reply:
Moreover, I think it would be slower since it's sn extra step if spans need to be used to select text later or when they are called seperately, while spacy already returns both in the same iterator

Reply:
no strong preference from my side. Don't think I'll use those offsets though.

Reply:
Ok so currently I chose to return list of tuples with text & start indice for both sent as well as words. Mainly because 1) I think it is more efficient and less error prone to return offset and text in one call, and 2) at this point there seems to be no clear preference for iterator over list, so I decided it to not change it.